pkg/activator/handler/handler_test.go

```diff
@@ -268,6 +308,159 @@ func TestActivationHandler(t *testing.T) {
+}
+
+// Make sure if one breaker is overflowed, the requests to other revisions are still served
+func TestActivationHandler_OverflowSeveralRevisions(t *testing.T) {
+	wantedSuccess := 40
+	wantedFailure := 2
+	overallRequests := 42
+	revisions := 2
+	respCh := make(chan *httptest.ResponseRecorder, overallRequests)
+
+	breakerParams := queue.BreakerParams{QueueDepth: 10, MaxConcurrency: 10, InitialCapacity: 10}
+
+	throttler := getThrottler(breakerParams, t)
+
+	server := httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			io.WriteString(w, wantBody)
+		}),
+	)
+
+	for rev := 0; rev < revisions; rev++ {
+		config := Config{
+			namespace:   fmt.Sprintf("real-namespace-%d", rev),
+			revName:     "real-name",
+			serviceName: "real-name-service",
+			wantBody:    wantBody,
+			requests:    21,
+			respCh:      respCh,
+			revisions:   revisions,
+		}
+		act := newStubActivator(config.namespace, config.revName, server)
+		handler := getHandler(throttler, t, act)
+		sendRequests(config, handler)
+	}
+	assertResponses(wantedSuccess, wantedFailure, overallRequests, respCh, t)
+}
+
+type Config struct {
+	namespace     string
+	revName       string
+	serviceName   string
+	requests      int
+	respCh        chan *httptest.ResponseRecorder
+	wantBody      string
+	breakerParams queue.BreakerParams
+	revisions     int
+}
+
+func sendRequests(config Config, handler ActivationHandler) {
+	for i := 0; i < config.requests; i++ {
+		go func() {
+			resp := httptest.NewRecorder()
+			req := httptest.NewRequest("POST", "http://example.com", nil)
+			req.Header.Set(activator.RevisionHeaderNamespace, config.namespace)
+			req.Header.Set(activator.RevisionHeaderName, config.revName)
+			handler.ServeHTTP(resp, req)
+			config.respCh <- resp
+		}()
+	}
+}
+
+func getThrottler(breakerParams queue.BreakerParams, t *testing.T) *activator.Throttler {
+	endpointsGetter := func(activator.RevisionID) (int32, error) {
+		return breakerParams.InitialCapacity, nil
+	}
+	throttlerParams := activator.ThrottlerParams{BreakerParams: breakerParams, Logger: TestLogger(t), GetRevision: stubRevisionGetter, GetEndpoints: endpointsGetter}
+	throttler := activator.NewThrottler(throttlerParams)
+	return throttler
+}
+
+func run(conf Config, throttler *activator.Throttler, t *testing.T) {
+	server := httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			io.WriteString(w, conf.wantBody)
+		}),
+	)
+	act := newStubActivator(conf.namespace, conf.revName, server)
+	handler := getHandler(throttler, t, act)
+	sendRequests(conf, handler)
+}
+
+func getHandler(throttler *activator.Throttler, t *testing.T, act activator.Activator) ActivationHandler {
+	rt := util.RoundTripperFunc(func(r *http.Request) (*http.Response, error) {
+		resp, err := http.DefaultTransport.RoundTrip(r)
+		if err != nil {
+			return resp, err
+		}
+		time.Sleep(50 * time.Millisecond)
```

Comment:
Why is there a sleep here?

Reply:
When we test sending 21 requests we have only 20 available slots, if the requests are too fast, the 21st could get a free slot from one of the "speedy" requests. Thus the sleep to make sure the last request is always rejected. 
I added a comment.

Reply:
Can we do this in a channel or a wait group of sorts? Sleeps usually generate lots of flakes.

Reply:
Do you have an exact place where you would have placed them?

Please consider my comment above, I'm planning to refactor the tests in the next PRs. It would be way easier to do that once the throttler is separated from the activating handler.

Reply:
If you're going to refactor them -- sure, we can leave it until later.